recipe-server/normandy/recipes/models.py

```diff
@@ -517,19 +550,55 @@ def recipes_used_by(self):
     def __str__(self):
         return self.name
 
+    def canonical_json(self):
+        # Avoid circular import
+        from normandy.recipes.api.v1.serializers import ActionSerializer
+        data = ActionSerializer(self).data
+        return CanonicalJSONRenderer().render(data)
+
     def get_absolute_url(self):
         return reverse('action-detail', args=[self.name])
 
     def compute_implementation_hash(self):
         return hashlib.sha1(self.implementation.encode()).hexdigest()
 
+    def update_signature(self):
+        try:
+            autographer = Autographer()
+        except ImproperlyConfigured:
+            self.signature = None
+            return
+
+        logger.info(
+            f'Requesting signatures for action named {self.name} from Autograph',
+            extra={'code': INFO_REQUESTING_ACTION_SIGNATURES, 'action_names': [self.name]}
+        )
+
+        signature_data = autographer.sign_data([self.canonical_json()])[0]
+        signature = Signature(**signature_data)
+        signature.save()
+        self.signature = signature
```

Comment:
I'm still mildly grumpy that this method saves the signature without saving this object it assigns the signature to, but I also still don't know a great alternative. Man.